ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java

```diff
@@ -417,6 +413,293 @@ Map<String, DataSource> getDataSources() {
+      statement.setInt(1, accountId);
+      statement.setInt(2, containerId);
+      statement.setString(3, record.getBlobName());
+      statement.setBytes(4, Base64.decodeBase64(record.getBlobId()));
+      if (record.getExpirationTimeMs() != Utils.Infinite_Time) {
+        statement.setTimestamp(5, new Timestamp(record.getExpirationTimeMs()));
+      } else {
+        statement.setTimestamp(5, null);
+      }
+      statement.executeUpdate();
+    } catch (SQLIntegrityConstraintViolationException e) {
+      rowAlreadyExists = true;
+    }
+    // 2. If the row already exists, attempt an update, checking that the update should be allowed (the current
+    //    row in the db represents a blob that was soft deleted or expired. Since soft deleted records are not
+    //    deleted from the table before a long retention period, there is no risk of the row not existing when
+    //    this query is run.
+    if (rowAlreadyExists) {
+      try (PreparedStatement statement = connection.prepareStatement(UPDATE_IF_DELETED_OR_EXPIRED_QUERY)) {
+        // fields to update
+        statement.setBytes(1, Base64.decodeBase64(record.getBlobId()));
+        if (record.getExpirationTimeMs() != Utils.Infinite_Time) {
+          statement.setTimestamp(2, new Timestamp(record.getExpirationTimeMs()));
+        } else {
+          statement.setTimestamp(2, null);
+        }
+        // primary key fields
+        statement.setInt(3, accountId);
+        statement.setInt(4, containerId);
+        statement.setString(5, record.getBlobName());
+        // the number of rows found will be 0 if there is currently an entry for this blob name that is
+        // not expired or deleted.
+        if (statement.executeUpdate() == 0) {
+          throw buildException("PUT: Blob still alive", RestServiceErrorCode.Conflict, record.getAccountName(),
+              record.getContainerName(), record.getBlobName());
+        }
+      }
+    }
+    return new PutResult(record);
+  }
+
+  private PutResult run_put_v2(NamedBlobRecord record, short accountId, short containerId, Connection connection)
+      throws Exception {
+    // 1. Attempt to insert into the table. This is attempted first since it is the most common case.
+    try (PreparedStatement statement = connection.prepareStatement(INSERT_QUERY_V2)) {
+      statement.setInt(1, accountId);
+      statement.setInt(2, containerId);
+      statement.setString(3, record.getBlobName());
+      statement.setBytes(4, Base64.decodeBase64(record.getBlobId()));
+      if (record.getExpirationTimeMs() != Utils.Infinite_Time) {
+        statement.setTimestamp(5, new Timestamp(record.getExpirationTimeMs()));
+      } else {
+        statement.setTimestamp(5, null);
+      }
+      statement.setLong(6, buildVersion());
+      statement.setInt(7, NamedBlobState.READY.ordinal());
+      statement.executeUpdate();
+    } catch (SQLIntegrityConstraintViolationException e) {
+      throw e;
```

Comment:
let's call build exception here.

Reply:
The issue is that the 'buildException' method need 'accountName', 'containerName', 'blobName':
```
  private static RestServiceException buildException(String message, RestServiceErrorCode errorCode, String accountName,
      String containerName, String blobName) {
```
While we only have 'accountId', 'containerId' at above line. To call build exception, we need refactoring of the methods signatures.

Plus that any exception raised here will be caught in the 'put' method for logging/metrics track purpose only, it won't be exposed to client, so I don't see benefits of call build exception here.
```
          try {
            run_put_v2(record, accountId, containerId, connection);
          } catch (Exception e) {
            logger.error("NAMED PUT: Data Error: old table succeed while new table failed: ", e);
            metricsRecoder.namedDataErrorPutCount.inc();
          }
```

Reply:
if that's the case let's not catch this exception, there is no point to just catch an exception and then throw it right away without doing anything.

Reply:
Done in latest commit.